File: MangaViewer/app/src/main/java/com/king/mangaviewer/activity/MainActivity.java
package com.king.mangaviewer.activity;

import android.app.SearchManager;
import android.content.Context;
import android.content.Intent;
import android.content.res.Configuration;
import android.content.res.TypedArray;
import android.os.Bundle;
import android.os.Handler;
import android.support.design.internal.NavigationMenuView;
import android.support.design.widget.NavigationView;
import android.support.design.widget.TabLayout;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentPagerAdapter;
import android.support.v4.view.ViewPager;
import android.support.v7.app.ActionBarDrawerToggle;
import android.support.v4.app.FragmentManager;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.widget.SearchView;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.WindowManager;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.ListView;
import android.widget.PopupWindow;
import android.widget.Toast;

import com.king.mangaviewer.R;
import com.king.mangaviewer.adapter.NavDrawerListAdapter;
import com.king.mangaviewer.model.NavDrawerItem;
import com.king.mangaviewer.model.MangaWebSource;
import com.king.mangaviewer.service.AutoNotifyUpdatedService;

import java.util.ArrayList;
import java.util.List;

public class MainActivity extends BaseActivity {

    private static final long DELAYTIME = 5000;
    private static final String STATE_KEY_POSITION = "state_key_position";
    CharSequence mTitle;
    CharSequence mDrawerTitle;
    private DrawerLayout mDrawerLayout;
    private TabLayout mTabLayout;
    private ListView mDrawerList;
    private ActionBarDrawerToggle mDrawerToggle;
    private ViewPagerAdapter mViewPagerAdapter;
    private SearchView searchView;
    private BaseFragment fragment = null;
    // slide menu items
    private String[] navMenuTitles;
    private TypedArray navMenuIcons;
    private ArrayList<NavDrawerItem> navDrawerItems;
    private NavigationView navigationView;
    private ViewPager mViewPager;
    private int mSelectedPosition;
    private int mTwoTapToExit;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        initDrawer(savedInstanceState);
    }

    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);
        //Log.i("MainActivity", "OnNewIntent");
        if (intent.getBooleanExtra(AutoNotifyUpdatedService.AUTO_UPDATE_SERVICE, false)) {
            intent.putExtra(AutoNotifyUpdatedService.AUTO_UPDATE_SERVICE, false);
//            displayView(getResources().getInteger(R.integer.menu_favourite_pos));
        }
    }

    @Override
    protected void onResume() {
        super.onResume();
        //Log.i("MainActivity", "onResume");
        if (searchView != null) {
            searchView.onActionViewCollapsed();
        }
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        outState.putInt(STATE_KEY_POSITION, mSelectedPosition);
        super.onSaveInstanceState(outState);
    }

    private void initDrawer(Bundle savedInstanceState) {
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        navigationView = (NavigationView) findViewById(R.id.navigation_view);
        navigationView.setNavigationItemSelectedListener(new NavigationView.OnNavigationItemSelectedListener() {
            @Override
            public boolean onNavigationItemSelected(MenuItem item) {

                switch (item.getItemId()) {
                    case R.id.menu_latest_manga:
                        mViewPager.setCurrentItem(0);
                        break;
                    case R.id.menu_all_manga:
                        mViewPager.setCurrentItem(1);
                        break;
                    case R.id.menu_favorite:
                        mViewPager.setCurrentItem(2);
                        break;
                    case R.id.menu_history:
                        mViewPager.setCurrentItem(3);
                        break;
                    case R.id.menu_local:
                        mViewPager.setCurrentItem(4);
                        break;
                    case R.id.menu_all_settings:
                        startActivity(new Intent(MainActivity.this, SettingsActivity.class));
                        break;
                    default:
                        break;
                }
                mDrawerLayout.closeDrawers();
                return true;
            }
        });

        mDrawerLayout = (DrawerLayout) findViewById(R.id.drawer_layout);
        mTitle = mDrawerTitle = getTitle();
        mDrawerToggle = new ActionBarDrawerToggle(this, mDrawerLayout,
                R.string.app_name, // nav drawer open - description for accessibility
                R.string.app_name // nav drawer close - description for accessibility
        ) {
            public void onDrawerClosed(View view) {
                getSupportActionBar().setTitle(mTitle);
                // calling onPrepareOptionsMenu() to show action bar icons
                invalidateOptionsMenu();
            }

            public void onDrawerOpened(View drawerView) {
                getSupportActionBar().setTitle(mDrawerTitle);
                // calling onPrepareOptionsMenu() to hide action bar icons
                invalidateOptionsMenu();
            }
        };
        mDrawerLayout.setDrawerListener(mDrawerToggle);

//        if (savedInstanceState == null) {
//            // on first time display view for first nav item
//            if (getIntent().getBooleanExtra(AutoNotifyUpdatedService.AUTO_UPDATE_SERVICE, false)) {
//                getIntent().putExtra(AutoNotifyUpdatedService.AUTO_UPDATE_SERVICE, false);
//                displayView(getResources().getInteger(R.integer.menu_favourite_pos));
//            } else {
//                displayView(0);
//            }
//        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
//        displayView(mSelectedPosition);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.main_menu, menu);
        // Associate searchable configuration with the SearchView
        SearchManager searchManager =
                (SearchManager) getSystemService(Context.SEARCH_SERVICE);
        searchView =
                (SearchView) menu.findItem(R.id.menu_search).getActionView();
        searchView.setSearchableInfo(
                searchManager.getSearchableInfo(getComponentName()));
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (mDrawerToggle.onOptionsItemSelected(item)) {
            return true;
        }
        switch (item.getItemId()) {
            case R.id.menu_setting:
                View v = findViewById(R.id.menu_setting);
                displayMangaSource(v);
                return true;

            default:
                return super.onOptionsItemSelected(item);
        }
    }

    private void displayMangaSource(View anchorView) {
        final PopupWindow popup = new PopupWindow(this);
        View layout = getLayoutInflater().inflate(R.layout.menu_main_setting, null);

        ListView lv = (ListView) layout.findViewById(R.id.listView);

        final List<MangaWebSource> mws = getAppViewModel().Setting.getMangaWebSources();
        int tSelectWebSourcePos = 0;
        List<String> source = new ArrayList<>();
        if (mws != null) {
            int i = 0;
            for (MangaWebSource m : mws) {
                source.add(m.getDisplayName());
                if (m.getId() == getAppViewModel().Setting.getSelectedWebSource(this).getId()) {
                    tSelectWebSourcePos = i;
                }
                i++;
            }
        }

        lv.setAdapter(new ArrayAdapter<String>(this, android.R.layout.simple_list_item_single_choice, source));
        lv.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                getAppViewModel().Setting.setSelectedWebSource(mws.get(position), MainActivity.this);
                getAppViewModel().Manga.setMangaMenuList(null);
                int currPos = mViewPager.getCurrentItem();
                mViewPager.setAdapter(mViewPagerAdapter);
                mViewPager.setCurrentItem(currPos);
//                ((BaseFragment)mViewPagerAdapter.getItem(mViewPager.getCurrentItem())).refresh();
                popup.dismiss();
            }
        });
        lv.setItemChecked(tSelectWebSourcePos, true);
        popup.setContentView(layout);
        // Set content width and height
        popup.setHeight(WindowManager.LayoutParams.WRAP_CONTENT);
        popup.setWidth(WindowManager.LayoutParams.WRAP_CONTENT);
        // Closes the popup window when touch outside of it - when looses focus
        popup.setOutsideTouchable(true);
        popup.setFocusable(true);
        // Show anchored to button
        popup.showAsDropDown(anchorView);
    }

    /**
     * When using the ActionBarDrawerToggle, you must call it during
     * onPostCreate() and onConfigurationChanged()...
     */
    @Override
    protected void onPostCreate(Bundle savedInstanceState) {
        super.onPostCreate(savedInstanceState);
        // Sync the toggle state after onRestoreInstanceState has occurred.
        mDrawerToggle.syncState();
    }

    @Override
    public void onConfigurationChanged(Configuration newConfig) {
        super.onConfigurationChanged(newConfig);
        // Pass any configuration change to the drawer toggls
        mDrawerToggle.onConfigurationChanged(newConfig);
    }

    @Override
    protected void initControl() {
        // TODO Auto-generated method stub
        setContentView(R.layout.activity_main_menu);
        mTabLayout = (TabLayout) findViewById(R.id.tabLayout);
        mViewPager = (ViewPager) findViewById(R.id.viewPager);
        setupViewPager(mViewPager);
        mTabLayout.setupWithViewPager(mViewPager);
        mTwoTapToExit = 0;
    }

    private void setupViewPager(ViewPager viewPager) {
        mViewPagerAdapter = new ViewPagerAdapter(getSupportFragmentManager());

        mViewPagerAdapter.addFragment(new HomeFragment(), getString(R.string.nav_latest_update));
        mViewPagerAdapter.addFragment(new AllMangaFragment(), getString(R.string.nav_all_manga));
        mViewPagerAdapter.addFragment(new FavouriteFragment(), getString(R.string.nav_favourite));
        mViewPagerAdapter.addFragment(new HistoryFragment(), getString(R.string.nav_history));
        mViewPagerAdapter.addFragment(new LocalFragment(), getString(R.string.nav_local));
        viewPager.setAdapter(mViewPagerAdapter);
    }

    @Override
    public void onBackPressed() {
        if (mTwoTapToExit < 1) {
            if (!mDrawerLayout.isDrawerOpen(navigationView)) {
                mDrawerLayout.openDrawer(navigationView);
            } else {
                mDrawerLayout.closeDrawer(navigationView);
            }
            mTwoTapToExit++;
            exitAppHandler.removeCallbacks(exitAppRunable);
            exitAppHandler.postDelayed(exitAppRunable, DELAYTIME);
            Toast.makeText(this, getString(R.string.msg_tap_two_to_exit), Toast.LENGTH_LONG).show();
        } else {
            super.onBackPressed();
        }

    }

    private Handler exitAppHandler = new Handler();
    private Runnable exitAppRunable = new Runnable() {
        @Override
        public void run() {
            mTwoTapToExit = 0;
        }
    };

    class ViewPagerAdapter extends FragmentPagerAdapter {
        private final List<Fragment> mFragmentList = new ArrayList<>();
        private final List<String> mFragmentTitleList = new ArrayList<>();

        public ViewPagerAdapter(FragmentManager manager) {
            super(manager);
        }

        @Override
        public Fragment getItem(int position) {
            return mFragmentList.get(position);
        }

        @Override
        public int getCount() {
            return mFragmentList.size();
        }

        public void addFragment(Fragment fragment, String title) {
            mFragmentList.add(fragment);
            mFragmentTitleList.add(title);
        }

        @Override
        public CharSequence getPageTitle(int position) {
            return mFragmentTitleList.get(position);
        }
    }
}

File: MangaViewer/app/src/main/java/com/king/mangaviewer/ui/main/MainActivity.kt
package com.king.mangaviewer.ui.main

import android.app.SearchManager
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.res.Configuration
import android.os.Bundle
import android.os.Handler
import android.support.design.widget.NavigationView
import android.support.design.widget.TabLayout
import android.support.design.widget.TabLayout.OnTabSelectedListener
import android.support.design.widget.TabLayout.Tab
import android.support.v4.app.Fragment
import android.support.v4.app.FragmentManager
import android.support.v4.app.FragmentStatePagerAdapter
import android.support.v4.view.ViewPager
import android.support.v4.widget.DrawerLayout
import android.support.v7.app.ActionBarDrawerToggle
import android.support.v7.widget.SearchView
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.WindowManager
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.ListView
import android.widget.PopupWindow
import android.widget.Toast
import com.king.mangaviewer.R
import com.king.mangaviewer.base.BaseActivity
import com.king.mangaviewer.base.ViewModelFactory
import com.king.mangaviewer.di.annotation.ActivityScopedFactory
import com.king.mangaviewer.service.AutoUpdateAlarmReceiver
import com.king.mangaviewer.ui.main.fragment.AllMangaFragment
import com.king.mangaviewer.ui.main.fragment.FavouriteFragment
import com.king.mangaviewer.ui.main.fragment.HistoryFragment
import com.king.mangaviewer.ui.main.fragment.HomeFragment
import com.king.mangaviewer.ui.main.fragment.LocalFragment
import com.king.mangaviewer.ui.search.SearchResultActivity
import com.king.mangaviewer.ui.setting.SettingsActivity
import com.king.mangaviewer.util.AppNavigator
import kotlinx.android.synthetic.main.activity_main_menu.fab
import kotlinx.android.synthetic.main.activity_main_menu.tabLayout
import java.util.ArrayList
import javax.inject.Inject

class MainActivity : BaseActivity() {
    private val FAVOURITE_POS = 0
    private val ALL_POS = 1

    internal var mTitle: CharSequence = ""
    internal var mDrawerTitle: CharSequence = ""
    lateinit var mDrawerLayout: DrawerLayout
    lateinit var mTabLayout: TabLayout
    lateinit var mDrawerToggle: ActionBarDrawerToggle
    lateinit var mViewPagerAdapter: ViewPagerAdapter
    private var searchView: SearchView? = null
    private var fragment: Fragment? = null
    // slide menu items
    lateinit var navigationView: NavigationView
    lateinit var mViewPager: ViewPager
    private val mSelectedPosition: Int = 0
    private var mTwoTapToExit: Int = 0

    @Inject
    @field:ActivityScopedFactory
    lateinit var viewModelFactory: ViewModelFactory

    @Inject
    lateinit var appNavigator: AppNavigator

    private val exitAppHandler = Handler()
    private val exitAppRunable = Runnable { mTwoTapToExit = 0 }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        initDrawer(savedInstanceState)
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        //Log.i("MainActivity", "OnNewIntent");
        if (intent.getBooleanExtra(AutoUpdateAlarmReceiver.AUTO_UPDATE_SERVICE,
                        false)) {
            intent.putExtra(AutoUpdateAlarmReceiver.AUTO_UPDATE_SERVICE, false)
            //            displayView(getResources().getInteger(R.integer.menu_favourite_pos));
            mViewPager.currentItem = FAVOURITE_POS
        }
    }

    override fun IsCanBack(): Boolean {
        return false
    }

    override fun onResume() {
        super.onResume()
        //Log.i("MainActivity", "onResume");
        searchView?.onActionViewCollapsed()

    }

    override fun onSaveInstanceState(outState: Bundle) {
        outState.putInt(STATE_KEY_POSITION, mSelectedPosition)
        super.onSaveInstanceState(outState)
    }

    private fun initDrawer(savedInstanceState: Bundle?) {
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)

        navigationView = findViewById<View>(R.id.navigation_view) as NavigationView
        navigationView.setNavigationItemSelectedListener { item ->
            when (item.itemId) {
                R.id.menu_latest_manga -> mViewPager.currentItem = 0
                R.id.menu_all_manga -> mViewPager.currentItem = 1
                R.id.menu_favorite -> mViewPager.currentItem = 2
                R.id.menu_history -> mViewPager.currentItem = 3
                R.id.menu_local -> mViewPager.currentItem = 4
                R.id.menu_all_settings -> startActivity(
                        Intent(this@MainActivity, SettingsActivity::class.java))
                else -> {
                }
            }

            mDrawerLayout.closeDrawers()
            true
        }

        mDrawerLayout = findViewById<View>(R.id.drawer_layout) as DrawerLayout
        mDrawerTitle = title
        mTitle = mDrawerTitle
        mDrawerToggle = object : ActionBarDrawerToggle(this, mDrawerLayout,
                R.string.app_name, // nav drawer open - description for accessibility
                R.string.app_name // nav drawer close - description for accessibility
        ) {
            override fun onDrawerClosed(view: View) {
                supportActionBar!!.setTitle(mTitle)
                // calling onPrepareOptionsMenu() to show action bar icons
                invalidateOptionsMenu()
            }

            override fun onDrawerOpened(drawerView: View) {
                supportActionBar!!.setTitle(mDrawerTitle)
                // calling onPrepareOptionsMenu() to hide action bar icons
                invalidateOptionsMenu()
            }
        }
        mDrawerLayout.setDrawerListener(mDrawerToggle)

        if (savedInstanceState == null) {
            // on first time display view for first nav item
            if (intent.getBooleanExtra(
                            AutoUpdateAlarmReceiver.AUTO_UPDATE_SERVICE, false)) {
                intent.putExtra(AutoUpdateAlarmReceiver.AUTO_UPDATE_SERVICE,
                        false)
                mViewPager.currentItem = ALL_POS
            } else {
                //
            }
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.main_menu, menu)
        // Associate searchable configuration with the SearchView
        val searchManager = getSystemService(Context.SEARCH_SERVICE) as SearchManager
        searchView = menu.findItem(R.id.menu_search).actionView as SearchView
        searchView?.setSearchableInfo(
                searchManager.getSearchableInfo(
                        ComponentName(this, SearchResultActivity::class.java)))
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (mDrawerToggle.onOptionsItemSelected(item)) {
            return true
        }
        return when (item.itemId) {
            R.id.menu_setting -> {
                val v = findViewById<View>(R.id.menu_setting)
                displayMangaSource(v)
                true
            }

            else -> super.onOptionsItemSelected(item)
        }
    }

    private fun displayMangaSource(anchorView: View) {
        val popup = PopupWindow(this)
        val layout = layoutInflater.inflate(R.layout.menu_main_setting, null)

        val lv = layout.findViewById<View>(R.id.listView) as ListView

        val mws = appViewModel.Setting.mangaWebSources
        var tSelectWebSourcePos = 0
        val source = ArrayList<String>()
        if (mws != null) {
            for ((i, m) in mws.withIndex()) {
                source.add(m.displayName)
                if (m.id == appViewModel.Setting.selectedWebSource.id) {
                    tSelectWebSourcePos = i
                }
            }
        }

        lv.adapter = ArrayAdapter(this, android.R.layout.simple_list_item_single_choice,
                source)
        lv.onItemClickListener = AdapterView.OnItemClickListener { parent, view, position, id ->
            appViewModel.Setting.setSelectedWebSource(mws!![position],
                    this@MainActivity)
            appViewModel.Manga.mangaMenuList = null
            val currPos = mViewPager.currentItem
            mViewPager.adapter = mViewPagerAdapter
            mViewPager.currentItem = currPos
            //                ((BaseFragment)mViewPagerAdapter.getItem(mViewPager.getCurrentItem())).refresh();
            popup.dismiss()
        }
        lv.setItemChecked(tSelectWebSourcePos, true)
        popup.contentView = layout
        // Set content width and height
        popup.height = WindowManager.LayoutParams.WRAP_CONTENT
        popup.width = WindowManager.LayoutParams.WRAP_CONTENT
        // Closes the popup window when touch outside of it - when looses focus
        popup.isOutsideTouchable = true
        popup.isFocusable = true
        // Show anchored to button
        popup.showAsDropDown(anchorView)
    }

    /**
     * When using the ActionBarDrawerToggle, you must call it during
     * onPostCreate() and onConfigurationChanged()...
     */
    override fun onPostCreate(savedInstanceState: Bundle?) {
        super.onPostCreate(savedInstanceState)
        // Sync the toggle state after onRestoreInstanceState has occurred.
        mDrawerToggle.syncState()
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
        // Pass any configuration change to the drawer toggls
        mDrawerToggle.onConfigurationChanged(newConfig)
    }

    override fun initControl() {
        setContentView(R.layout.activity_main_menu)
        mTabLayout = findViewById<View>(R.id.tabLayout) as TabLayout
        mViewPager = findViewById<View>(R.id.viewPager) as ViewPager
        setupViewPager(mViewPager)
        tabLayout.addOnTabSelectedListener(object : OnTabSelectedListener {
            override fun onTabReselected(tab: Tab?) {
                val pos = mViewPager.currentItem
                fragment = mViewPagerAdapter.getItem(pos)
                (fragment as? HasFloatActionButton)?.initFab(fab)
            }

            override fun onTabUnselected(tab: Tab?) {
            }

            override fun onTabSelected(tab: Tab?) {
                val pos = mViewPager.currentItem
                fragment = mViewPagerAdapter.getItem(pos)
                if (fragment is HasFloatActionButton) {
                    (fragment as HasFloatActionButton).initFab(fab)
                } else {
                    fab.hide()
                }
            }

        })

        fab.setOnClickListener {
            (fragment as? HasFloatActionButton)?.onClick()
        }

        mTabLayout.setupWithViewPager(mViewPager)
        mTwoTapToExit = 0
    }

    private fun setupViewPager(viewPager: ViewPager) {
        mViewPagerAdapter = ViewPagerAdapter(supportFragmentManager)

        mViewPagerAdapter.addFragment(FavouriteFragment(), getString(R.string.nav_favourite))
        mViewPagerAdapter.addFragment(HomeFragment(), getString(R.string.nav_latest_update))
        mViewPagerAdapter.addFragment(AllMangaFragment(), getString(R.string.nav_all_manga))
        mViewPagerAdapter.addFragment(HistoryFragment(), getString(R.string.nav_history))
        mViewPagerAdapter.addFragment(LocalFragment(), getString(R.string.nav_local))

        viewPager.adapter = mViewPagerAdapter
        viewPager.currentItem = ALL_POS
    }

    override fun onBackPressed() {
        if (mTwoTapToExit < 1) {
            if (!mDrawerLayout.isDrawerOpen(navigationView)) {
                mDrawerLayout.openDrawer(navigationView)
            } else {
                mDrawerLayout.closeDrawer(navigationView)
            }
            mTwoTapToExit++
            exitAppHandler.removeCallbacks(exitAppRunable)
            exitAppHandler.postDelayed(exitAppRunable, DELAYTIME)
            Toast.makeText(this, getString(R.string.msg_tap_two_to_exit), Toast.LENGTH_LONG).show()
        } else {
            super.onBackPressed()
        }

    }

    class ViewPagerAdapter(manager: FragmentManager) :
            FragmentStatePagerAdapter(manager) {
        private val mFragmentList = ArrayList<Fragment>()
        private val mFragmentTitleList = ArrayList<String>()

        override fun getItem(position: Int): Fragment {
            return mFragmentList[position]
        }

        override fun getCount(): Int {
            return mFragmentList.size
        }

        fun addFragment(fragment: Fragment, title: String) {
            mFragmentList.add(fragment)
            mFragmentTitleList.add(title)
        }

        override fun getPageTitle(position: Int): CharSequence? {
            return mFragmentTitleList[position]
        }

    }

    companion object {

        private val DELAYTIME: Long = 5000
        private val STATE_KEY_POSITION = "state_key_position"
    }
}
